### DISCUSSION TITLE
How can I include Manufacturer Data in BLE advertisements?

### DISCUSSION THREAD

**Discussion by DavesCodeMusings** [micropython/micropython#15921] [Libraries & Drivers]
I've put together a proof of concept DIY Bluetooth Low Energy beacon using an ESP32 and the Arduino C++ platform. What it does is to communicate a battery level in the _manufacturer data_ portion of the BLE advertisement.

I'd like to do the same with MicroPython and aioble.

The idea is to broadcast data at periodic intervals rather than requiring clients to connect and poll a GATT characteristic. That way the ESP32 can spend most of its time in deep sleep, waking up just long enough to send out a few dozen advertisements.

It works great with Arduino BLE. Here's the key part of the code:

```
byte data[3] = { 0xA3, 0x09, (byte) batteryPercent};
BLE.stopAdvertise();  // Need to re-advertise new data, so stop, update, and start.
BLE.setManufacturerData(data, 3);
BLE.advertise();
```

The `data[]` array contains two bytes for the manufacturer's ID (at the mement, I'm using 0x09A3 which is reistered to Arduino SA.) The last byte is the level of the battery charge between 0 and 100.

The problem I'm running across is I cannot find the MicroPython equivalent of `BLE.setManufacturerData()`

Any suggestions as to how I can set this _manufacturer data_ portion of the BLE Advertisement with MicroPython and aioble?

**Comment by walklinewang** [comment]
Here's the way to set `manufacturer data`, but it's not an API way, you have to modify your code, for example:

* Define a constant in [examples/bluetooth/ble_advertising.py](https://github.com/micropython/micropython/blob/master/examples/bluetooth/ble_advertising.py)

    ```python
    _ADV_TYPE_MANUFACTURER_SPECIFIC_DATA = const(0xFF)
    ```

* In the same file, append a new line in [advertising_payload](https://github.com/micropython/micropython/blob/17d82344581ad3a76033fae54c5d3304e17f185f/examples/bluetooth/ble_advertising.py#L30) function.

    ```python
    _append(_ADV_TYPE_MANUFACTURER_SPECIFIC_DATA, pack('<2H', 0x09a3, 0x03))
    ```

* And then you will got these results
    ![89317a2325e26d9cc0539400a6c82eb](https://github.com/user-attachments/assets/85ff3263-3de8-45da-b629-6d79109d395b)
    ![9b821c8d1119950947c5a52179d286d](https://github.com/user-attachments/assets/df6b43e1-c806-4ae1-bc1a-97129169966c)

Please ignore the duplicated items in these screenshots.

**Comment by DavesCodeMusings** [comment]
With the suggestion from @Walkline80 in mind, I started looking into advertising with aioble. I found this example of a [peripheral server](https://github.com/micropython/micropython-lib/tree/master/micropython/bluetooth/aioble#register-services-and-wait-for-connection-peripheral-server)

I'm wondering if this bit might be what I'm looking for:

```
while True:
    connection = await aioble.advertise(
            _ADV_INTERVAL_US,
            name="temp-sense",
            services=[_ENV_SENSE_UUID],
            appearance=_GENERIC_THERMOMETER,
            manufacturer=(0xabcd, b"1234"),
        )
```

The line with `manufacturer=(0xabcd, b"1234"),` looks suspiciously like the format used to set manufacturer data: a 16-bit manufacturer ID followed by some arbitrary amount of byte data.

**Comment by DavesCodeMusings** [comment]
I got it to work by taking the aioble [temperature sensor example](https://github.com/micropython/micropython-lib/blob/master/micropython/bluetooth/aioble/examples/temp_sensor.py) and adding in `manufacturer=(0xabcd, b"1234")`.

So this...

```py
async def peripheral_task():
    while True:
        async with await aioble.advertise(
            _ADV_INTERVAL_MS,
            name="mpy-temp",
            services=[_ENV_SENSE_UUID],
            appearance=_ADV_APPEARANCE_GENERIC_THERMOMETER,
        ) as connection:
            print("Connection from", connection.device)
            await connection.disconnected(timeout_ms=None)
```

Became this...

```py
async def peripheral_task():
    while True:
        async with await aioble.advertise(
            _ADV_INTERVAL_MS,
            name="mpy-temp",
            services=[_ENV_SENSE_UUID],
            appearance=_ADV_APPEARANCE_GENERIC_THERMOMETER,
            manufacturer=(0xabcd, b"1234")
        ) as connection:
            print("Connection from", connection.device)
            await connection.disconnected(timeout_ms=None)
```

And bingo! Manufacturer data showed up in the scan.

![ble_screenshot](https://github.com/user-attachments/assets/f1bfaef8-5928-4c64-b4dd-0c2cf62b590a)

**Comment by bidrohini** [comment]
Have you installed the aioble library from here? https://github.com/micropython/micropython-lib/blob/master/micropython/bluetooth/aioble/README.md

**Reply by DavesCodeMusings** [reply]
Yes. I actually got this working with aioble. I have [the code on my GitHub site](https://github.com/DavesCodeMusings/BLESP32), along with a short write up of the process and how to send the advertised readings into Home Assistant using ESPHome's Bluetooth gateway.
